src/evict/evict_lru.c

```diff
@@ -541,42 +538,49 @@ __evict_update_work(WT_SESSION_IMPL *session)
 	cache->evict_max_refs_per_file =
 	    WT_MAX(100, WT_MILLION / (conn->open_file_count + 1));
 
+	if (cache->evict_queues[WT_EVICT_URGENT_QUEUE].evict_current != NULL)
+		FLD_SET(cache->state, WT_EVICT_STATE_URGENT);
+
 	/*
-	 * Page eviction overrides the dirty target and other types of eviction,
-	 * that is, we don't care where we are with respect to the dirty target
-	 * if page eviction is configured.
+	 * If we need space in the cache, try to find clean pages to evict.
 	 *
 	 * Avoid division by zero if the cache size has not yet been set in a
 	 * shared cache.
 	 */
 	bytes_max = conn->cache_size + 1;
 	bytes_inuse = __wt_cache_bytes_inuse(cache);
-	if (bytes_inuse > (cache->eviction_target * bytes_max) / 100) {
-		FLD_SET(cache->state, WT_EVICT_STATE_ALL);
-		goto done;
-	}
+	if (bytes_inuse > (cache->eviction_target * bytes_max) / 100)
+		FLD_SET(cache->state, WT_EVICT_STATE_CLEAN);
 
 	/*
-	 * If the cache has been stuck and is now under control, clear the
-	 * stuck flag.
+	 * Scrub dirty pages and keep them in cache if we are less than half
+	 * way between the cache target and trigger.
 	 */
-	if (bytes_inuse < bytes_max)
-		F_CLR(cache, WT_CACHE_STUCK);
+	if (bytes_inuse < ((cache->eviction_target + cache->eviction_trigger) *
+	    bytes_max) / 200)
+		FLD_SET(cache->state, WT_EVICT_STATE_SCRUB);
```

Comment:
This policy has very little hard data behind it -- I think it's reasonable, and could be revisited later: it came out of a discussion rather than any formal design.